File: ios/ui-framework-sample/UIFramework/Sources/UIFramework/BizKhmerKeyboard.swift
import SwiftUI

public struct BizKhmerKeyboard: View {
    public var onPress: (String) -> Void
    
    @State private var isShifted = false
    @State private var showSymbols = false
    
    @State private var shieldPositions: [Int] = [0, 0, 0, 0]
    
    let khmerRow0 = ["១","២","៣","៤","៥","៦","៧","៨","៩","០"]
    let khmerRow0Shift = ["!","@","#","$","%","^","&","*","(",")"]
    
    let khmerDefaultLetters = ["ឆ","វ","េ","រ","ត","យ","ុ","ិ","ោ","ផ","ា","ស","ឌ","ថ","ង","ហ","្","ក","ល","ឋ","ខ","ច","វ","ប","ន","ម"]
    let khmerShiftedLetters = ["ឈ","ឺ","ែ","ឫ","ទ","យ","ូ","ី","ៅ","ភ","ាំ","ស្ស","ធ","ញ","ង","ះ","្រ","គ","ឡ","ឍ","ឃ","ជ","ឈ","ព","ណ","ំ"]
    
    let defaultSymbols = ["!","@","#","$","%","^","&","*","(",")","-","+","_","=","{","}","[","]",":",";","\"","'","<",">","?","/"]
    
    public init(onPress: @escaping (String) -> Void) {
        self.onPress = onPress
    }
    
    public var body: some View {
        VStack(spacing: 8) {
            // Row 1
            HStack(spacing: 6) {
                let source = isShifted ? khmerRow0Shift : khmerRow0
                ForEach(0..<11, id: \.self) { index in
                    if index == shieldPositions[0] {
                        ShieldIndicator()
                    } else {
                        let arrayIndex = index > shieldPositions[0] ? index - 1 : index
                        let char = source[arrayIndex]
                        KeyboardKey(text: char, isShifted: false, action: { emitPress(char) })
                    }
                }
            }
            
            // Row 2
            HStack(spacing: 6) {
                let source = showSymbols ? defaultSymbols : (isShifted ? khmerShiftedLetters : khmerDefaultLetters)
                ForEach(0..<11, id: \.self) { index in
                    if index == shieldPositions[1] {
                        ShieldIndicator()
                    } else {
                        let arrayIndex = index > shieldPositions[1] ? index - 1 : index
                        let char = source[arrayIndex]
                        KeyboardKey(text: char, isShifted: false, action: { emitPress(char) })
                    }
                }
            }
            
            // Row 3
            HStack(spacing: 6) {
                let source = showSymbols ? defaultSymbols : (isShifted ? khmerShiftedLetters : khmerDefaultLetters)
                ForEach(0..<10, id: \.self) { index in
                    if index == shieldPositions[2] {
                        ShieldIndicator()
                    } else {
                        let arrayIndex = (index > shieldPositions[2] ? index - 1 : index) + 10
                        let char = source[arrayIndex]
                        KeyboardKey(text: char, isShifted: false, action: { emitPress(char) })
                    }
                }
            }
            .padding(.horizontal, 16)
            
            // Row 4
            HStack(spacing: 6) {
                Button(action: toggleShift) {
                    ZStack {
                        RoundedRectangle(cornerRadius: 8)
                            .fill(Color(hex: "#202124"))
                        
                        Image(systemName: isShifted ? "shift.fill" : "shift")
                            .foregroundColor(isShifted ? Color(hex: "#3880FF") : Color(hex: "#E8EAED"))
                    }
                }
                .frame(maxWidth: 54, maxHeight: .infinity)
                
                let source = showSymbols ? defaultSymbols : (isShifted ? khmerShiftedLetters : khmerDefaultLetters)
                ForEach(0..<8, id: \.self) { index in
                    if index == shieldPositions[3] {
                        ShieldIndicator()
                    } else {
                        let arrayIndex = (index > shieldPositions[3] ? index - 1 : index) + 19
                        if arrayIndex < source.count {
                            let char = source[arrayIndex]
                            KeyboardKey(text: char, isShifted: false, action: { emitPress(char) })
                        }
                    }
                }
                
                Button(action: { emitPress("backspace") }) {
                    ZStack {
                        RoundedRectangle(cornerRadius: 8)
                            .fill(Color(hex: "#202124"))
                        
                        Image(systemName: "delete.left")
                            .foregroundColor(Color(hex: "#E8EAED"))
                    }
                }
                .frame(maxWidth: 54, maxHeight: .infinity)
            }
            .frame(height: 48)
            
            // Row 5
            HStack(spacing: 6) {
                Button(action: toggleSymbols) {
                    ZStack {
                        RoundedRectangle(cornerRadius: 8)
                            .fill(Color(hex: "#202124"))
                        
                        Text(showSymbols ? "កខគ" : "$%?")
                            .font(.system(size: 16, weight: .semibold))
                            .foregroundColor(Color(hex: "#E8EAED"))
                    }
                }
                .frame(maxWidth: 64, maxHeight: .infinity)
                
                Button(action: { emitPress("emoji") }) {
                    ZStack {
                        RoundedRectangle(cornerRadius: 8)
                            .fill(Color(hex: "#202124"))
                        Text("😊")
                    }
                }
                .frame(maxWidth: 54, maxHeight: .infinity)
                
                Button(action: { emitPress(" ") }) {
                    RoundedRectangle(cornerRadius: 8)
                        .fill(Color(hex: "#2C2D31"))
                }
                .frame(maxWidth: .infinity, maxHeight: .infinity)
                
                Button(action: { emitPress(".") }) {
                    ZStack {
                        RoundedRectangle(cornerRadius: 8)
                            .fill(Color(hex: "#202124"))
                        Text(".")
                            .font(.system(size: 20, weight: .medium))
                            .foregroundColor(Color(hex: "#E8EAED"))
                    }
                }
                .frame(maxWidth: 54, maxHeight: .infinity)
                
                Button(action: { emitPress("enter") }) {
                    ZStack {
                        RoundedRectangle(cornerRadius: 8)
                            .fill(Color(hex: "#202124"))
                        
                        Image(systemName: "return")
                            .foregroundColor(Color(hex: "#E8EAED"))
                    }
                }
                .frame(maxWidth: 54, maxHeight: .infinity)
            }
            .frame(height: 48)
        }
        .padding(EdgeInsets(top: 8, leading: 4, bottom: 24, trailing: 4))
        .background(Color(hex: "#1A1B1E"))
        .onAppear {
            randomizeKeys()
        }
    }
    
    private func toggleShift() {
        isShifted.toggle()
        if showSymbols {
            showSymbols = false
        }
    }
    
    private func toggleSymbols() {
        showSymbols.toggle()
        if showSymbols {
            isShifted = false
        }
    }
    
    private func emitPress(_ val: String) {
        onPress(val)
    }
    
    public func randomizeKeys() {
        shieldPositions = [
            Int.random(in: 0..<11),
            Int.random(in: 0..<11),
            Int.random(in: 0..<10),
            Int.random(in: 0..<8)
        ]
    }
}

private struct ShieldIndicator: View {
    var body: some View {
        ZStack {
            Color.clear
            Image(systemName: "shield.lefthalf.filled")
                .foregroundColor(Color(hex: "#5F6368"))
                .font(.system(size: 16))
        }
        .frame(maxWidth: 40, maxHeight: .infinity)
    }
}

private struct KeyboardKey: View {
    let text: String
    let isShifted: Bool
    let action: () -> Void
    
    var body: some View {
        Button(action: action) {
            ZStack {
                RoundedRectangle(cornerRadius: 8)
                    .fill(Color(hex: "#2C2D31"))
                    .shadow(color: Color.black.opacity(0.2), radius: 1, x: 0, y: 1)
                
                Text(text)
                    .font(.system(size: 20, weight: .medium))
                    .foregroundColor(.white)
            }
        }
        .frame(maxWidth: 40, maxHeight: 48)
    }
}

File: android/ui-framework-sample/ui-framework/src/main/java/kh/mcnc/lib/components/BizKhmerKeyboard.kt
package kh.mcnc.lib.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

@Composable
fun BizKhmerKeyboard(
    modifier: Modifier = Modifier,
    onPress: (String) -> Unit
) {
    var isShifted by remember { mutableStateOf(false) }
    var showSymbols by remember { mutableStateOf(false) }

    var shieldPositions by remember { mutableStateOf(listOf(0, 0, 0, 0)) }

    val randomizeKeys = {
        shieldPositions = listOf(
            (0..10).random(),
            (0..10).random(),
            (0..9).random(),
            (0..7).random()
        )
    }

    LaunchedEffect(Unit) {
        randomizeKeys()
    }

    val khmerRow0 = listOf("១","២","៣","៤","៥","៦","៧","៨","៩","០")
    val khmerRow0Shift = listOf("!","@","#","$","%","^","&","*","(",")")

    val khmerDefaultLetters = listOf("ឆ","វ","េ","រ","ត","យ","ុ","ិ","ោ","ផ","ា","ស","ឌ","ថ","ង","ហ","្","ក","ល","ឋ","ខ","ច","វ","ប","ន","ម")
    val khmerShiftedLetters = listOf("ឈ","ឺ","ែ","ឫ","ទ","យ","ូ","ី","ៅ","ភ","ាំ","ស្ស","ធ","ញ","ង","ះ","្រ","គ","ឡ","ឍ","ឃ","ជ","ឈ","ព","ណ","ំ")

    val defaultSymbols = listOf("!","@","#","$","%","^","&","*","(",")","-","+","_","=","{","}","[","]",":",";","\"","'","<",">","?","/")

    val randomizedNumbers = remember(shieldPositions, isShifted) {
        val nums = if (isShifted) khmerRow0Shift.toMutableList() else khmerRow0.toMutableList()
        nums.add(shieldPositions[0], "shield")
        nums
    }

    val row1Keys = remember(shieldPositions, showSymbols, isShifted) {
        val source = if (showSymbols) defaultSymbols else if (isShifted) khmerShiftedLetters else khmerDefaultLetters
        val keys = source.take(10).toMutableList()
        keys.add(shieldPositions[1], "shield")
        keys
    }

    val row2Keys = remember(shieldPositions, showSymbols, isShifted) {
        val source = if (showSymbols) defaultSymbols else if (isShifted) khmerShiftedLetters else khmerDefaultLetters
        val keys = source.drop(10).take(9).toMutableList()
        keys.add(shieldPositions[2], "shield")
        keys
    }

    val row3Keys = remember(shieldPositions, showSymbols, isShifted) {
        val source = if (showSymbols) defaultSymbols else if (isShifted) khmerShiftedLetters else khmerDefaultLetters
        val keys = source.drop(19).take(7).toMutableList()
        keys.add(shieldPositions[3], "shield")
        keys
    }

    Column(
        modifier = modifier
            .fillMaxWidth()
            .background(Color(0xFF1A1B1E))
            .padding(top = 8.dp, start = 4.dp, end = 4.dp, bottom = 24.dp)
    ) {
        // Row 1
        Row(
            modifier = Modifier.fillMaxWidth().padding(bottom = 8.dp),
            horizontalArrangement = Arrangement.spacedBy(6.dp, Alignment.CenterHorizontally)
        ) {
            randomizedNumbers.forEach { item ->
                if (item == "shield") {
                    KhmerShieldKeyboardIndicator()
                } else {
                    KhmerKeyboardCharKey(
                        char = item,
                        onClick = { onPress(item) }
                    )
                }
            }
        }

        // Row 2
        Row(
            modifier = Modifier.fillMaxWidth().padding(bottom = 8.dp),
            horizontalArrangement = Arrangement.spacedBy(6.dp, Alignment.CenterHorizontally)
        ) {
            row1Keys.forEach { item ->
                if (item == "shield") {
                    KhmerShieldKeyboardIndicator()
                } else {
                    KhmerKeyboardCharKey(
                        char = item,
                        onClick = { onPress(item) }
                    )
                }
            }
        }

        // Row 3
        Row(
            modifier = Modifier.fillMaxWidth().padding(bottom = 8.dp, start = 16.dp, end = 16.dp),
            horizontalArrangement = Arrangement.spacedBy(6.dp, Alignment.CenterHorizontally)
        ) {
            row2Keys.forEach { item ->
                if (item == "shield") {
                    KhmerShieldKeyboardIndicator()
                } else {
                    KhmerKeyboardCharKey(
                        char = item,
                        onClick = { onPress(item) }
                    )
                }
            }
        }

        // Row 4
        Row(
            modifier = Modifier.fillMaxWidth().padding(bottom = 8.dp),
            horizontalArrangement = Arrangement.spacedBy(6.dp, Alignment.CenterHorizontally)
        ) {
            KhmerKeyboardFuncKey(
                content = { ShiftIconCanvas(isShifted = isShifted) },
                onClick = { 
                    isShifted = !isShifted
                    if (showSymbols) showSymbols = false
                }
            )

            row3Keys.forEach { item ->
                if (item == "shield") {
                    KhmerShieldKeyboardIndicator()
                } else {
                    KhmerKeyboardCharKey(
                        char = item,
                        onClick = { onPress(item) }
                    )
                }
            }

            KhmerKeyboardFuncKey(
                content = { BackspaceIconCanvas(tint = Color.White) },
                onClick = { onPress("backspace") }
            )
        }

        // Row 5
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(6.dp, Alignment.CenterHorizontally)
        ) {
            KhmerKeyboardFuncKey(
                modifier = Modifier.weight(1.5f),
                text = if (showSymbols) "ABC" else "$%?",
                onClick = {
                    showSymbols = !showSymbols
                    if (showSymbols) isShifted = false
                }
            )

            KhmerKeyboardFuncKey(
                modifier = Modifier.weight(1.5f),
                text = "😊",
                onClick = { onPress("emoji") }
            )

            // Space
            Box(
                modifier = Modifier
                    .weight(5f)
                    .height(48.dp)
                    .clip(RoundedCornerShape(8.dp))
                    .background(Color(0xFF2C2D31))
                    .clickable { onPress(" ") }
            )

            KhmerKeyboardFuncKey(
                modifier = Modifier.weight(1.5f),
                text = ".",
                onClick = { onPress(".") }
            )

            KhmerKeyboardFuncKey(
                modifier = Modifier.weight(1.5f),
                content = { EnterIconCanvas() },
                onClick = { onPress("enter") }
            )
        }
    }
}

@Composable
private fun KhmerKeyboardCharKey(
    modifier: Modifier = Modifier,
    char: String,
    onClick: () -> Unit
) {
    Box(
        modifier = modifier
            .widthIn(max = 40.dp)
            .height(48.dp)
            .shadow(1.dp, RoundedCornerShape(8.dp), spotColor = Color.Black.copy(alpha = 0.2f))
            .clip(RoundedCornerShape(8.dp))
            .background(Color(0xFF2C2D31))
            .clickable(onClick = onClick)
            .then(Modifier.wrapContentWidth()),
        contentAlignment = Alignment.Center
    ) {
        Text(
            text = char,
            fontSize = 20.sp,
            fontWeight = FontWeight.Medium,
            color = Color.White,
            modifier = Modifier.padding(horizontal = 8.dp)
        )
    }
}

@Composable
private fun KhmerKeyboardFuncKey(
    modifier: Modifier = Modifier,
    text: String? = null,
    content: (@Composable () -> Unit)? = null,
    onClick: () -> Unit
) {
    Box(
        modifier = modifier
            .widthIn(max = 50.dp)
            .height(48.dp)
            .clip(RoundedCornerShape(8.dp))
            .background(Color(0xFF202124))
            .clickable(onClick = onClick)
            .then(Modifier.wrapContentWidth()),
        contentAlignment = Alignment.Center
    ) {
        if (text != null) {
            Text(
                text = text,
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
                color = Color(0xFFAEB0B4),
                modifier = Modifier.padding(horizontal = 8.dp)
            )
        } else if (content != null) {
            content()
        }
    }
}

@Composable
private fun KhmerShieldKeyboardIndicator(modifier: Modifier = Modifier) {
    Box(
        modifier = modifier
            .widthIn(max = 40.dp)
            .height(48.dp),
        contentAlignment = Alignment.Center
    ) {
        Canvas(modifier = Modifier.size(20.dp)) {
            val w = size.width
            val h = size.height
            val path = Path().apply {
                moveTo(w * 0.5f, h * 0.1f)
                lineTo(w * 0.9f, h * 0.25f)
                lineTo(w * 0.9f, h * 0.6f)
                cubicTo(w * 0.9f, h * 0.8f, w * 0.5f, h * 0.9f, w * 0.5f, h * 0.9f)
                cubicTo(w * 0.5f, h * 0.9f, w * 0.1f, h * 0.8f, w * 0.1f, h * 0.6f)
                lineTo(w * 0.1f, h * 0.25f)
                close()
            }
            drawPath(
                path = path,
                color = Color(0xFF5F6368),
                style = Stroke(width = 1.5f.dp.toPx(), join = StrokeJoin.Round)
            )
        }
    }
}
